src/node_file.cc

```diff
@@ -2072,6 +2072,131 @@ static void ReadDir(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+static void ReaddirRecursiveSync(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+
+  CHECK_GE(args.Length(), 3);
+
+  BufferValue path(isolate, args[0]);
+  CHECK_NOT_NULL(*path);
+  ToNamespacedPath(env, &path);
+
+  const enum encoding encoding = ParseEncoding(isolate, args[1], UTF8);
+  bool with_file_types = args[2]->IsTrue();
+
+  THROW_IF_INSUFFICIENT_PERMISSIONS(
+      env, permission::PermissionScope::kFileSystemRead, path.ToStringView());
+
+  std::vector<std::string> paths;
+  std::vector<std::string> names;
+  std::vector<int> types;
+  std::vector<std::string> relative_paths;
+
+  std::vector<std::string> dir_queue;
+  dir_queue.push_back(*path);
+
+  FSReqWrapSync req_wrap_sync("readdir_recursive");
+
+  const std::string base_path(*path);
+
+  while (!dir_queue.empty()) {
+    std::string current_path = std::move(dir_queue.back());
+    dir_queue.pop_back();
+
+    int err = uv_fs_scandir(nullptr, &req_wrap_sync.req,
+                           current_path.c_str(), 0, nullptr);
+    if (err < 0) {
+      if (err == UV_ENOENT || err == UV_ENOTDIR) {
+        continue;
+      }
+      uv_fs_req_cleanup(&req_wrap_sync.req);
+      args.GetReturnValue().Set(v8::Undefined(isolate));
+      return;
+    }
+
+    uv_dirent_t ent;
+    while (uv_fs_scandir_next(&req_wrap_sync.req, &ent) != UV_EOF) {
+      if (strcmp(ent.name, ".") == 0 || strcmp(ent.name, "..") == 0) {
+        continue;
+      }
+
+      std::string full_path = current_path;
+      if (full_path.back() != kPathSeparator) {
+        full_path += kPathSeparator;
+      }
+      full_path += ent.name;
+
+      if (with_file_types) {
+        paths.push_back(current_path);
+        names.emplace_back(ent.name);
+        types.push_back(static_cast<int>(ent.type));
+      } else {
+        if (full_path.length() > base_path.length()) {
+          size_t start_pos = base_path.length();
+          if (full_path[start_pos] == kPathSeparator) {
+            start_pos++;
+          }
+          if (start_pos < full_path.length()) {
+            relative_paths.emplace_back(full_path.substr(start_pos));
+          }
+        }
+      }
+
+      if (ent.type == UV_DIRENT_DIR) {
+        dir_queue.push_back(std::move(full_path));
+      }
+    }
+
+    uv_fs_req_cleanup(&req_wrap_sync.req);
+  }
+
+  Local<Array> result_array = Array::New(isolate,
```

Comment:
creating an std::array, mutating them after the for loops below, and later assigning it to a Local<Array>  might make the code more readable (and faster)